udpsession.go

```diff
@@ -0,0 +1,364 @@
+	return "Tuna Session UDP"
+}
+func (usa *udpSessionAddr) String() string {
+	return sessionKey(usa.remoteAddr, usa.sessionID)
+}
+
+// UDP session, it is a UDP listener or dialer endpoint for reading and writing udp data.
+// After tuna session set up UDP connection, it return UpdSession to applications.
+// Applications use this end point to read and write data.
+// If there is a node break down between UDP listener and dialer, udp session will reconenct automatically
+// without interrupting application usage.
+type UdpSession struct {
+	endpointType int // EndpointDialer or EndpointListener
+	udpConn      *tuna.EncryptUDPConn
+	ts           *TunaSessionClient // Reference of tuna session
+
+	// dialer only
+	tcpConn *Conn           // TCP *Conn, dialer only
+	peer    *udpSessionAddr // dialer only
+
+	// listener only
+	sync.RWMutex
+	sessKeyToSessAddr map[string]*udpSessionAddr // listener only, udpSessionAddr.String() to *udpSessionAddr
+	udpAddrToSessAddr map[string]*udpSessionAddr // listener only, net.UDPAddr.String() to *udpSessionAddr
+	tcpConns          map[string]*Conn           // listen only, session key to *Conn
+}
+
+func newUdpSession(ts *TunaSessionClient, udpConn *tuna.EncryptUDPConn, endpointType int, peerAddr *udpSessionAddr) *UdpSession {
+	us := &UdpSession{ts: ts, udpConn: udpConn, endpointType: endpointType, peer: peerAddr}
+	if endpointType == EndpointListener {
+		us.sessKeyToSessAddr = make(map[string]*udpSessionAddr)
+		us.udpAddrToSessAddr = make(map[string]*udpSessionAddr)
+		if peerAddr != nil {
+			us.sessKeyToSessAddr[peerAddr.String()] = peerAddr
+			us.udpAddrToSessAddr[peerAddr.udpAddr.String()] = peerAddr
+		}
+		us.tcpConns = make(map[string]*Conn)
+	}
+
+	return us
+}
+
+// ReadFrom, implement of standard library net.UDPConn.ReadFrom
+func (us *UdpSession) ReadFrom(b []byte) (int, net.Addr, error) {
+	for {
+		buf := make([]byte, tuna.MaxUDPBufferSize)
+		n, udpAddr, err := us.udpConn.ReadFromUDP(buf)
+		if err != nil {
+			return 0, nil, err
+		}
+
+		header := buf[0]
+		switch header {
+		case byte(pb.HeaderType_USER):
+			us.RLock()
+			addr, ok := us.udpAddrToSessAddr[udpAddr.String()]
+			us.RUnlock()
+			if ok {
+				copy(b, buf[1:n]) // remove header, and return to user application
+				return n - 1, addr, err
+			} else {
+				log.Printf("UdpSession.ReadFrom addr %v which is not in map, but got user data len %v\n", udpAddr, n)
+			}
+
+		case byte(pb.HeaderType_SESSION):
+			err := us.handleSessionMsg(buf[:n], udpAddr)
+			if err != nil {
+				return 0, nil, err
+			}
+
+		default:
+			return 0, nil, ErrWrongMsgFormat
+		}
+	}
+}
+
+func (us *UdpSession) Read(b []byte) (int, error) {
+	n, _, err := us.ReadFrom(b)
+	return n, err
+}
+
+// ReadFrom, implement of standard library net.UDPConn.WriteTo
+// Only for user data write
+func (us *UdpSession) WriteTo(b []byte, addr net.Addr) (n int, err error) {
+	buf := make([]byte, 1+len(b))
+	buf[0] = byte(pb.HeaderType_USER)
+	copy(buf[1:], b)

```

Comment:
Need to encrypt message. The easiest way is to call AddCodec after creating udpConn

Reply:
Yes, added.